### PR TITLE
CBMC: Add spec+proof for `gen_matrix_entry()`

### LINKED ISSUE #326: State and prove spec for gen_matrix_entry

### PR DESCRIPTION
**Pull request by hanno-becker**
This was pretty hard. I did not manage to verify `gen_matrix_entry()` in acceptable time as it stood, because the reasoning about the pointer increment in the coefficient buffer somehow tripped up the SAT solvers. Instead, I had to modify the signature of `rej_uniform()` to always take the polynomial base -- luckily, this arguably makes the code cleaner anyway. With that signature change in place, things are quite fast. See commit messages for some detail.

Resolves #326

### CODE REVIEW

File: mlkem/rej_uniform.c [hunk 1]
```diff
@@ -23,23 +25,24 @@
...
- * have been consumed.
+ * Returns the new offset of sampled 16-bit integers, at most target,
+ * and at least the initial offset.
+ * If the new offset is strictly less than len, all of the input buffers
+ * is guaranteed to have been consumed. If it is equal to len, no information
+ * is provided on how many bytes of the input buffer have been consumed.
  **************************************************/
-static unsigned int rej_uniform_scalar(int16_t *r, unsigned int len,
-                                       const uint8_t *buf,
+static unsigned int rej_uniform_scalar(int16_t *r, unsigned target,
```
Comment (mkannwischer) [inline mlkem/rej_uniform.c:36]: nit: `unsigned` vs `unsigned int`

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks @hanno-becker. 
The bounds look good to me and the refactoring compared to the ref implementation is acceptable. Small nits.

File: fips202/fips202.h [hunk 2]
```diff
@@ -40,14 +40,19 @@ typedef struct {
  * with the same state.
  */
 #define shake128_absorb FIPS202_NAMESPACE(shake128_absorb)
-void shake128_absorb(shake128ctx *state, const uint8_t *input, size_t inlen);
+void shake128_absorb(shake128ctx *state, const uint8_t *input, size_t inlen)
+    // TODO: Expand + prove spec
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply fips202/fips202.h:44]: Yes

File: mlkem/rej_uniform.c (hunk 1 above)
Comment (hanno-becker) [reply mlkem/rej_uniform.c:36]: fixing

**Review by mkannwischer — APPROVED** [review]
LGTM now. Thanks for the edits.

**Pull request merged**